File: apps/frontend/src/adapters/shared/ai-arena.ts
import type {
  ArenaAgent,
  ArenaChallenge,
  ArenaLeaderboard,
  ArenaParticipant,
  ArenaPortfolio,
  ArenaRun,
  ArenaTrade,
  CompanyThesis,
  CreateArenaAgentRequest,
  CreateArenaChallengeRequest,
  CreateCompanyThesisRequest,
  RunArenaAgentRequest,
} from "@/lib/types";
import { invoke, logger } from "./platform";

export const getArenaAgents = async (): Promise<ArenaAgent[]> => {
  try {
    return await invoke<ArenaAgent[]>("get_arena_agents");
  } catch (error) {
    logger.error("Error fetching arena agents.");
    throw error;
  }
};

export const createArenaAgent = async (agent: CreateArenaAgentRequest): Promise<ArenaAgent> => {
  try {
    return await invoke<ArenaAgent>("create_arena_agent", { agent });
  } catch (error) {
    logger.error("Error creating arena agent.");
    throw error;
  }
};

export const getArenaChallenges = async (): Promise<ArenaChallenge[]> => {
  try {
    return await invoke<ArenaChallenge[]>("get_arena_challenges");
  } catch (error) {
    logger.error("Error fetching arena challenges.");
    throw error;
  }
};

export const createArenaChallenge = async (
  challenge: CreateArenaChallengeRequest,
): Promise<ArenaChallenge> => {
  try {
    return await invoke<ArenaChallenge>("create_arena_challenge", { challenge });
  } catch (error) {
    logger.error("Error creating arena challenge.");
    throw error;
  }
};

export const joinArenaChallenge = async (
  challengeId: string,
  agentId: string,
): Promise<ArenaParticipant> => {
  try {
    return await invoke<ArenaParticipant>("join_arena_challenge", { challengeId, agentId });
  } catch (error) {
    logger.error("Error joining arena challenge.");
    throw error;
  }
};

export const getArenaParticipants = async (challengeId: string): Promise<ArenaParticipant[]> => {
  try {
    return await invoke<ArenaParticipant[]>("get_arena_participants", { challengeId });
  } catch (error) {
    logger.error("Error fetching arena participants.");
    throw error;
  }
};

export const runArenaAgent = async (request: RunArenaAgentRequest): Promise<ArenaRun> => {
  try {
    return await invoke<ArenaRun>("run_arena_agent", { request });
  } catch (error) {
    logger.error("Error running arena agent.");
    throw error;
  }
};

export const runDueArenaAgents = async (): Promise<ArenaRun[]> => {
  try {
    return await invoke<ArenaRun[]>("run_due_arena_agents");
  } catch (error) {
    logger.error("Error running due arena agents.");
    throw error;
  }
};

export const settleArenaChallenge = async (challengeId: string): Promise<ArenaLeaderboard> => {
  try {
    return await invoke<ArenaLeaderboard>("settle_arena_challenge", { challengeId });
  } catch (error) {
    logger.error("Error settling arena challenge.");
    throw error;
  }
};

export const getArenaLeaderboard = async (challengeId: string): Promise<ArenaLeaderboard> => {
  try {
    return await invoke<ArenaLeaderboard>("get_arena_leaderboard", { challengeId });
  } catch (error) {
    logger.error("Error fetching arena leaderboard.");
    throw error;
  }
};

export const getArenaPortfolio = async (participantId: string): Promise<ArenaPortfolio> => {
  try {
    return await invoke<ArenaPortfolio>("get_arena_portfolio", { participantId });
  } catch (error) {
    logger.error("Error fetching arena portfolio.");
    throw error;
  }
};

export const getArenaRuns = async (challengeId: string): Promise<ArenaRun[]> => {
  try {
    return await invoke<ArenaRun[]>("get_arena_runs", { challengeId });
  } catch (error) {
    logger.error("Error fetching arena runs.");
    throw error;
  }
};

export const getArenaTrades = async (challengeId: string): Promise<ArenaTrade[]> => {
  try {
    return await invoke<ArenaTrade[]>("get_arena_trades", { challengeId });
  } catch (error) {
    logger.error("Error fetching arena trades.");
    throw error;
  }
};

export const createCompanyThesis = async (
  thesis: CreateCompanyThesisRequest,
): Promise<CompanyThesis> => {
  try {
    return await invoke<CompanyThesis>("create_company_thesis", { thesis });
  } catch (error) {
    logger.error("Error creating company thesis.");
    throw error;
  }
};

export const getCompanyTheses = async (input?: {
  symbol?: string;
  challengeId?: string;
  limit?: number;
}): Promise<CompanyThesis[]> => {
  try {
    return await invoke<CompanyThesis[]>("get_company_theses", input ?? {});
  } catch (error) {
    logger.error("Error fetching company theses.");
    throw error;
  }
};

File: apps/tauri/src/commands/ai_arena.rs
use std::sync::Arc;

use tauri::State;

use crate::context::ServiceContext;
use wealthfolio_core::ai_arena::{
    ArenaAgent, ArenaChallenge, ArenaLeaderboard, ArenaParticipant, ArenaPortfolio, ArenaRun,
    ArenaTrade, CompanyThesis, CreateArenaAgentRequest, CreateArenaChallengeRequest,
    CreateCompanyThesisRequest, RunArenaAgentRequest,
};

#[tauri::command]
pub async fn get_arena_agents(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ArenaAgent>, String> {
    state
        .ai_arena_service()
        .list_agents()
        .map_err(|e| format!("Failed to load arena agents: {}", e))
}

#[tauri::command]
pub async fn create_arena_agent(
    agent: CreateArenaAgentRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaAgent, String> {
    state
        .ai_arena_service()
        .create_agent(agent)
        .await
        .map_err(|e| format!("Failed to create arena agent: {}", e))
}

#[tauri::command]
pub async fn get_arena_challenges(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ArenaChallenge>, String> {
    state
        .ai_arena_service()
        .list_challenges()
        .map_err(|e| format!("Failed to load arena challenges: {}", e))
}

#[tauri::command]
pub async fn create_arena_challenge(
    challenge: CreateArenaChallengeRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaChallenge, String> {
    state
        .ai_arena_service()
        .create_challenge(challenge)
        .await
        .map_err(|e| format!("Failed to create arena challenge: {}", e))
}

#[tauri::command]
pub async fn join_arena_challenge(
    challenge_id: String,
    agent_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaParticipant, String> {
    state
        .ai_arena_service()
        .join_challenge(&challenge_id, &agent_id)
        .await
        .map_err(|e| format!("Failed to join arena challenge: {}", e))
}

#[tauri::command]
pub async fn get_arena_participants(
    challenge_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ArenaParticipant>, String> {
    state
        .ai_arena_service()
        .list_participants(&challenge_id)
        .map_err(|e| format!("Failed to load arena participants: {}", e))
}

#[tauri::command]
pub async fn run_arena_agent(
    request: RunArenaAgentRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaRun, String> {
    state
        .ai_arena_service()
        .run_agent(request)
        .await
        .map_err(|e| format!("Failed to run arena agent: {}", e))
}

#[tauri::command]
pub async fn run_due_arena_agents(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ArenaRun>, String> {
    state
        .ai_arena_service()
        .run_due_scheduled()
        .await
        .map_err(|e| format!("Failed to run due arena agents: {}", e))
}

#[tauri::command]
pub async fn settle_arena_challenge(
    challenge_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaLeaderboard, String> {
    state
        .ai_arena_service()
        .settle_challenge(&challenge_id)
        .await
        .map_err(|e| format!("Failed to settle arena challenge: {}", e))
}

#[tauri::command]
pub async fn get_arena_leaderboard(
    challenge_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaLeaderboard, String> {
    state
        .ai_arena_service()
        .get_leaderboard(&challenge_id)
        .await
        .map_err(|e| format!("Failed to load arena leaderboard: {}", e))
}

#[tauri::command]
pub async fn get_arena_portfolio(
    participant_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ArenaPortfolio, String> {
    state
        .ai_arena_service()
        .get_portfolio(&participant_id)
        .await
        .map_err(|e| format!("Failed to load arena portfolio: {}", e))
}

#[tauri::command]
pub async fn get_arena_runs(
    challenge_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ArenaRun>, String> {
    state
        .ai_arena_service()
        .list_runs(&challenge_id)
        .map_err(|e| format!("Failed to load arena runs: {}", e))
}

#[tauri::command]
pub async fn get_arena_trades(
    challenge_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ArenaTrade>, String> {
    state
        .ai_arena_service()
        .list_trades(&challenge_id)
        .map_err(|e| format!("Failed to load arena trades: {}", e))
}

#[tauri::command]
pub async fn create_company_thesis(
    thesis: CreateCompanyThesisRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<CompanyThesis, String> {
    state
        .ai_arena_service()
        .create_thesis(thesis)
        .await
        .map_err(|e| format!("Failed to create company thesis: {}", e))
}

#[tauri::command]
pub async fn get_company_theses(
    symbol: Option<String>,
    challenge_id: Option<String>,
    limit: Option<i64>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<CompanyThesis>, String> {
    state
        .ai_arena_service()
        .list_theses(
            symbol.as_deref(),
            challenge_id.as_deref(),
            limit.unwrap_or(50),
        )
        .map_err(|e| format!("Failed to load company theses: {}", e))
}
